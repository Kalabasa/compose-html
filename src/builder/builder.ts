import { compileFile } from "compiler/compiler";
import { Component } from "compiler/component";
import { toHTML } from "dom/dom";
import glob from "glob";
import { copyFileSync, lstatSync, mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";
import { Renderer } from "renderer/renderer";
import { html as beautifyHTML, HTMLBeautifyOptions } from "js-beautify";
import { createLogger } from "util/log";

type BuildOptions = {
  inputDir?: string;
  outputDir?: string;
  rootDir?: string;
  beautify?: HTMLBeautifyOptions | false;
};

const DEFAULT_OPTIONS = {
  inputDir: process.cwd(),
  outputDir: path.resolve(process.cwd(), "out"),
  beautify: {
    extra_liners: [],
  },
} satisfies BuildOptions;

const logger = createLogger(path.basename(__filename, ".ts"));

export function build(options: BuildOptions = {}) {
  const {
    inputDir,
    outputDir,
    rootDir: rootDirOption,
    beautify,
  } = Object.assign({}, DEFAULT_OPTIONS, options);
  const rootDir = rootDirOption ?? inputDir;

  logger.info(
    "✍  compose-html",
    "\n\nWorking directories",
    "\n   input:",
    formatPath(inputDir),
    "\n  output:",
    formatPath(outputDir),
    "\n    root:",
    formatPath(rootDir),
    "\n"
  );

  const nodir = { nodir: true };
  const htmlFiles = glob.sync(path.resolve(inputDir, "**/*.html"), nodir);
  const nonHTMLFiles = glob
    .sync(path.resolve(rootDir, "**/*"), nodir)
    .filter((f) => !htmlFiles.includes(f));
  logger.info(htmlFiles.length, "html files");
  logger.info(nonHTMLFiles.length, "non-html files");

  // compile HTML files as components
  const componentMap = new Map<string, Component>();
  for (const filePath of htmlFiles) {
    const component = compileFile(filePath);
    componentMap.set(component.name, component);
  }
  const renderer = new Renderer(componentMap);
  logger.debug("Loaded components:", componentMap.keys());
 
  // copy non-HTML files
  for (const file of nonHTMLFiles) {
    const relPath = path.relative(rootDir, file);
    const outFilePath = path.resolve(outputDir, relPath);
    mkdirSync(path.dirname(outFilePath), { recursive: true });
    copyFileSync(file, outFilePath);
    logger.info("Copied", formatPath(file), "→", formatPath(outFilePath));
  }

  // render pages
  for (const component of componentMap.values()) {
    if (!component.page || !component.filePath.startsWith(rootDir)) continue;

    let pageHTML = toHTML(renderer.render(component));

    if (beautify) {
      initBeautifyDefaults(beautify, pageHTML);
      pageHTML = beautifyHTML(pageHTML, beautify);
    }

    const pagePath = path.relative(rootDir, component.filePath);
    const outFilePath =
      component.name === "index"
        ? path.resolve(outputDir, pagePath)
        : path.resolve(
            outputDir,
            path.dirname(pagePath),
            component.name,
            "index.html"
          );

    mkdirSync(path.dirname(outFilePath), { recursive: true });
    writeFileSync(outFilePath, pageHTML);
    logger.info(
      "Rendered",
      formatPath(component.filePath),
      "→",
      formatPath(outFilePath)
    );
  }
}

function initBeautifyDefaults(
  beautify: { extra_liners: never[] } & HTMLBeautifyOptions,
  pageHTML: string
) {
  if (!beautify.indent_with_tabs && beautify.indent_size == undefined) {
    // detect indent
    const indentMatch = pageHTML.match(/\n(\s+)(?=\S)/);
    if (indentMatch) {
      const char = indentMatch[1][0];
      if (char === "\t") {
        beautify.indent_with_tabs = true;
      } else {
        beautify.indent_char = char;
        beautify.indent_size = indentMatch[1].length;
      }
    }
  }
}

function formatPath(p: string): string {
  const cwd = process.cwd();
  const abs = path.resolve(p);
  if (abs.startsWith(cwd)) {
    return path.relative(cwd, p);
  } else {
    return abs;
  }
}
