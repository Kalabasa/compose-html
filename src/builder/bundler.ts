import { Component } from "compiler/component";
import { createElement, isInlineJavaScriptElement } from "dom/dom";
import path from "node:path";
import { createLogger } from "util/log";
import { checkNotNull } from "util/preconditions";

export type Page = {
  pagePath: string;
  nodes: Node[];
};

type Bundle = {
  src: string;
  code: string;
};

const logger = createLogger(path.basename(__filename, ".ts"));

export function extractScriptBundles(
  pages: Page[],
  minPageUsage: number,
  srcPrefix: string,
  components: Iterable<Component>
): Array<Bundle> {
  const { scriptPages, scriptElements, scriptComponents } = mapScripts(
    pages,
    components
  );

  logger.debug(
    "script to pages map:\n" +
      Array.from(scriptPages.entries())
        .map(
          ([script, pages]) =>
            `  \`${script.replaceAll(/\s+/g, " ").slice(0, 80)}\`\n` +
            `      => ${pages.map((page) => page.pagePath).join(", ")}`
        )
        .join("\n")
  );

  const scriptBundles = generateBundles(
    scriptElements,
    scriptComponents,
    scriptPages,
    minPageUsage
  );

  mergeBundles(scriptPages, scriptElements, scriptBundles);

  const bundleSet = new Set(scriptBundles.values());

  for (const bundle of bundleSet) {
    bundle.src = `${srcPrefix}${bundle.src}.js`;
  }

  replaceScriptsWithBundles(pages, scriptBundles);

  return [...bundleSet];
}

// Merge bundles that are always included together in the same pages
function mergeBundles(
  scriptPages: Map<string, Page[]>,
  scriptElements: Map<string, HTMLScriptElement>,
  scriptBundles: Map<string, Bundle>
): void {
  // bucket key = script loading type + page usage signature / hash
  const buckets = new Map<string, Bundle>();

  for (const [scriptHTML, bundle] of scriptBundles.entries()) {
    const element = checkNotNull(scriptElements.get(scriptHTML));
    const loadingType =
      (element.hasAttribute("async") ? "a" : "s") +
      (element.hasAttribute("defer") ? "d" : "e");

    const pages = scriptPages.get(scriptHTML) ?? [];

    const signature =
      loadingType +
      "::" +
      pages
        .map((page) => page.pagePath)
        .sort()
        .join(":");

    const bucketBundle = buckets.get(signature);
    if (bucketBundle) {
      // merge bundles
      bucketBundle.src += "-" + bundle.src;
      bucketBundle.code += "\n" + bundle.code;
      // point to the same bundle
      scriptBundles.set(scriptHTML, bucketBundle);
    } else {
      buckets.set(signature, bundle);
    }
  }
}

function generateBundles(
  scriptElements: Map<string, HTMLScriptElement>,
  scriptComponents: Map<string, Component>,
  scriptPages: Map<string, Page[]>,
  minPageUsage: number
) {
  const scriptBundles = new Map<string, Bundle>();
  for (const [scriptHTML, pages] of scriptPages.entries()) {
    const element = checkNotNull(scriptElements.get(scriptHTML));

    const canBeExtracted =
      pages.length >= minPageUsage || element.hasAttribute("async");
    if (!canBeExtracted) continue;

    let src = "";
    const component = scriptComponents.get(scriptHTML);
    if (component) {
      src = component.name;
    } else {
      src = pages.map((page) => page.pagePath.replaceAll(/\W/g, "_")).join("-");
    }

    if (!src) continue;

    const bundle: Bundle = {
      src,
      code: element.innerHTML,
    };

    scriptBundles.set(scriptHTML, bundle);
  }
  return scriptBundles;
}

function replaceScriptsWithBundles(
  pages: Page[],
  scriptBundles: Map<string, Bundle>
) {
  for (const page of pages) {
    const included = new Set<Bundle>();

    for (const script of findScripts(page.nodes)) {
      const bundle = scriptBundles.get(script.outerHTML);
      if (bundle) {
        if (included.has(bundle)) {
          script.remove();
          continue;
        }

        included.add(bundle);

        const bundleScript = createElement("script");
        for (const attr of script.attributes) {
          bundleScript.setAttribute(attr.name, attr.value);
        }
        bundleScript.src = bundle.src;
        script.replaceWith(bundleScript);
      }
    }

    if (included.size) {
      logger.debug(
        `Extracting ${included.size} bundles from:`,
        page.pagePath,
        [...included].map((bundle) => "\n  " + bundle.src).join("")
      );
    }
  }
}

function mapScripts(pages: Page[], components: Iterable<Component>) {
  const scriptPages = new Map<string, Page[]>();
  const scriptElements = new Map<string, HTMLScriptElement>();
  const scriptComponents = new Map<string, Component>();

  for (const page of pages) {
    for (const script of findScripts(page.nodes)) {
      scriptElements.set(script.outerHTML, script);

      let pages = scriptPages.get(script.outerHTML);
      if (!pages) scriptPages.set(script.outerHTML, (pages = []));

      pages.push(page);
    }
  }

  for (const component of components) {
    for (const script of component.clientScripts) {
      scriptElements.set(script.outerHTML, script);
      scriptComponents.set(script.outerHTML, component);
    }
  }

  return { scriptPages, scriptElements, scriptComponents };
}

function* findScripts(nodes: Iterable<Node>): Generator<HTMLScriptElement> {
  for (const node of nodes) {
    if (isInlineJavaScriptElement(node)) yield node;
    else yield* findScripts(node.childNodes);
  }
}
