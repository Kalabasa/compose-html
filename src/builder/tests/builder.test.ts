import { build } from "builder/builder";
import { glob } from "glob";
import { lstatSync, mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { execSync } from "node:child_process";
import { relative, resolve } from "node:path";

describe("builder", () => {
  jest.setTimeout(20 * 1000);

  const topDir = resolve(__dirname, "../../..");
  const projects = ["project1", "project2"];
  const exclude = ["**/dont-process.html"];

  beforeAll(() =>
    inDir(topDir, () => {
      run("npm install");
    })
  );

  it.each(projects)("builds %s via api", async (projectDir) => {
    const inputDir = getInputDir(projectDir);
    await inDir(topDir, () =>
      withTempDir(async (outputDir) => {
        await build({
          inputDir,
          outputDir,
          exclude,
        });
        expect(filesToString(outputDir)).toMatchSnapshot();
      })
    );
  });

  it.each(projects)("builds %s via cli", async (projectDir) => {
    const inputDir = getInputDir(projectDir);
    await inDir(topDir, () =>
      withTempDir((outputDir) => {
        run(
          "node dist/bin/main.js" +
            ` -i ${relative(process.cwd(), inputDir)}` +
            ` -o ${outputDir}` +
            ` --exclude ${exclude.map((p) => `'${p}' `)}`
        );
        expect(filesToString(outputDir)).toMatchSnapshot();
      })
    );
  });

  it.each(projects)("builds %s via config file", async (projectDir) => {
    const inputDir = getInputDir(projectDir);
    await inDir(topDir, () =>
      withTempDir((outputDir) =>
        withTempFile(
          JSON.stringify({
            inputDir,
            outputDir,
            exclude,
          }),
          (configFile) => {
            run(`node dist/bin/main.js --config ${configFile}`);
            expect(filesToString(outputDir)).toMatchSnapshot();
          }
        )
      )
    );
  });

  it.each(projects)(
    "builds %s via config file with cli override",
    async (projectDir) => {
      const inputDir = getInputDir(projectDir);
      await inDir(topDir, () =>
        withTempDir((outputDir) =>
          withTempFile(
            JSON.stringify({
              inputDir,
              outputDir,
              exclude,
            }),
            (configFile) => {
              run(`node dist/bin/main.js --config ${configFile} -p about`);
              expect(filesToString(outputDir)).toMatchSnapshot();
            }
          )
        )
      );
    }
  );
});

function run(command: string) {
  console.log("> " + command);
  return execSync(command, { stdio: "inherit" });
}

async function inDir(dir: string, fn: Function) {
  const cwd = process.cwd();
  process.chdir(dir);
  try {
    await fn();
  } finally {
    process.chdir(cwd);
  }
}

async function withTempDir(
  fn: (dir: string) => void | Promise<void>
): Promise<void> {
  const tempDir = makeTempDir();
  try {
    await fn(tempDir);
  } finally {
    await rm(tempDir, { recursive: true, force: true });
  }
}

async function withTempFile(data: string, fn: (file: string) => void) {
  const file = resolve("jest-builder-tmpfile-" + Date.now().toString(36));
  try {
    writeFileSync(file, data);
    fn(file);
  } finally {
    await rm(file, { force: true });
  }
}

function getInputDir(projectDir: string) {
  return resolve(__dirname, "data", projectDir);
}

function makeTempDir() {
  return mkdtempSync(resolve(tmpdir(), "jest-builder-"));
}

function filesToString(rootDir: string): string {
  let result = "";

  const files = glob.sync(resolve(rootDir, "**"));
  for (const filePath of files) {
    if (!lstatSync(filePath).isFile()) continue;
    result += "./";
    result += relative(rootDir, filePath);
    result += ":\n";
    result += readFileSync(filePath);
    result += "\n\n";
  }

  return "\n" + result.trim() + "\n";
}
