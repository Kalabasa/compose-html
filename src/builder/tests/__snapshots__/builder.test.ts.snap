// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder builds project1 1`] = `
"./about/index.html:
<html>
<head>
  <title>About Project 1</title>
</head>
<body>
  <div class="header">
    <h1>About Project 1</h1>

  </div>
  <div>This is a test project.</div>
</body>
</html>

./index.html:
<html>
<head>
  <title>Project 1</title>
  <style>
    .button {
      border-radius: 11px;
    }
  </style>
</head>
<body>
  <div class="header">
    <h1>Project 1</h1>

    <button class="button">Sign in</button>
    <button class="button">Log up</button>

  </div>
  <div>Hello, world!</div>
  <nav>
    <a href="about.html">About</a>
  </nav>
</body>
</html>"
`;
